diff --git a/release/scripts/modules/addon_utils.py b/release/scripts/modules/addon_utils.py
index ef5ec77a6c5..2e2b61c4e54 100644
--- a/release/scripts/modules/addon_utils.py
+++ b/release/scripts/modules/addon_utils.py
@@ -38,12 +38,20 @@ error_duplicates = []
 addons_fake_modules = {}
 
 
+class Addon:
+    properties = ""
+    module = "ocvl-addon"
+
+ocvl = Addon()
+
+
 # called only once at startup, avoids calling 'reset_all', correct but slower.
 def _initialize():
     path_list = paths()
     for path in path_list:
         _bpy.utils._sys_path_ensure(path)
-    for addon in _preferences.addons:
+    addons_with_ocvl = list(_preferences.addons)  + [ocvl]
+    for addon in addons_with_ocvl:
         enable(addon.module)
 
 
